cornellh4i/greenzone
# Bug fixes

## Overview

This addresses issue #77. The backend for language translation has been completed, and a folder named language now contains the English and Mongolian `JSON` translations. When clicking on a county on the map, both the English and Mongolian names of the counties are displayed (see images below).

The sidebar, top bar page, and other relevant pages have been translated accurately into both English and Mongolian. Additionally, instructions for continuing the translation process and adapting it for other pages are provided below. The translations persist correctly, matching the toggle state as expected.

## Testing

I tested the implementation by adding new languages (English, Mongolian, Italian, Spanish), using console logs to track the changes in i18n, and verifying the updates directly on the website. Additionally, I conducted smaller-scale testing using another [repository](https://github.com/LiuBrianna/multiLangTest) to ensure the functionality worked as expected.

To test: 

1. Run the backend by `cd backend` then `yarn start`
2. Run the frontend by `cd frontend` then `yarn dev`

To test the side panel:
Run `http://localhost:3001/monitoring-platform` and interact with the toggle to switch between languages.
_Both the sidebar and top bar are implemented_ 

## Impact

GreenZone is designed to cater to both English and Mongolian users. To enhance accessibility and broaden its reach, we need to implement a language switcher, allowing users to easily toggle between the two languages.

## Screenshots

**Side Panel:**
<img width="1469" alt="image" src="https://github.com/user-attachments/assets/85f3660c-c41a-4933-bc24-3f73c5683a11">
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/48c794e4-6971-4f64-9334-a84ba7e3640c">
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/4cebabbc-55e1-4cb6-bfc2-8014ed389ed6">
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/89394f35-7c3c-411d-aa36-915a2fcf857f">

**Example of same county (Bulgan):** 
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/01071462-f05e-486e-b8da-9fd7201b54c8">
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/ef560d00-c7ac-47d9-8916-ee0c93084330">


**Top Panel:**
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/1859dc66-6826-42ce-9d7b-b5172590dc75">
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/0fa10a33-6978-4e5c-a230-1d9f4901ed07">
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/98601bf3-e716-4305-9213-9c8e1879d5b8">

**Page Routing:** 
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/8c0cc82f-4fd9-4799-85b1-7d071a5e5200">
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/f00f0f9b-6044-4a93-9d8e-ff942790da51">


## Notes

- Enhance the feature by utilizing the user's device settings to automatically detect and translate content.
- To accommodate a larger user base, we can extend language support beyond just English and Mongolian.
- Did not accommodate for fuzzy search for counties but it can be accounted since it is in the backend already
- When implementing a new page or feature, the translations should be added to the corresponding `json` files in the `backend/src/languages` and following the instructions below 

1. Add to the correspond json files
2. Import the below:
- `import { useTranslation } from "react-i18next";`
- `import "../multi-Lang/i18n"; // This will trigger i18n initialization`
- `import LangChange from "../multi-Lang/LangSwitch";` <-- this is for the toggle feature itself 
3. Call `const { t } = useTranslation();` because this where the languages are stored and will be able to update the languages as well.
4. Add this so the translation changes persist and re-render the components of the page or component
`useEffect(() => {
    // Any effect or side-effect you need when translations change
  }, [t]); // Re-run when t changes`
5. To call the translated text or sentence, call `t("<json key")` or `{t("<json key")}`
6. To render the toggle, call `<LangChange />`
<img width="199" alt="image" src="https://github.com/user-attachments/assets/7379614c-58db-4b1e-9576-0c06c894e6ef">
